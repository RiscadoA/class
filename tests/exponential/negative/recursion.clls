type rec Nat {
    choice of {
        |#Z: close
        |#S: Nat
    }
};;

proc rec printNat(n: ~Nat) {
    unfold n;
    case n of {
        |#Z: println("Z"); wait n; ()
        |#S: print("S "); printNat(n)
    }
};;

proc main() {
    cut {
        !n(n: ~Nat);
        printNat(n)
        |n: ?Nat|
        ?n;
        par {
            call n(x);
            unfold x; #S x; unfold x; #Z x; close x
            ||
            call n(x);
            unfold x; #S x; unfold x; #S x; unfold x; #Z x; close x
        }
    }
};;
