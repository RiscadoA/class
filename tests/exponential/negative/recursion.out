S Z
S S Z
